src/field_theory/separable_degree.lean

```diff
@@ -0,0 +1,179 @@
+- `separable_degree_dvd_degree'`: the separable degree divides the degree, in function of the
+  exponential characteristic of the field
+- `separable_degree_dvd_degree` and `separable_degree_eq_degree` specialize the statement of
+  `separable_degree_dvd_degree`
+- `separable_degree_eq`: the separable degree is unique (for polynomials over a field)
+
+## Tags
+
+separable degree, degree, polynomial
+-/
+
+namespace polynomial
+
+noncomputable theory
+open_locale classical
+
+section comm_semiring
+
+variables {F : Type} [comm_semiring F]
+  (q : ℕ)
+
+/-- A separable contraction of a polynomial `f` is a separable polynomial `g` such that
+`g(x^(q^m)) = f(x)` for some `m : ℕ`.-/
+def has_separable_contraction (f : polynomial F) : Prop :=
+∃ g : polynomial F, g.separable ∧ ∃ m : ℕ, expand F (q^m) g = f
+
+variables {f : polynomial F} (hf : has_separable_contraction q f)
+
+/-- A choice of a separable contraction. -/
+def separable_contraction : polynomial F := classical.some hf
+
+/-- The separable degree of a polynomial is the degree of a given separable contraction. -/
+def separable_degree : ℕ := nat_degree (separable_contraction q hf)
+
+/-- The separable degree divides the degree, in function of the exponential characteristic of F. -/
+lemma separable_degree_dvd_degree' : ∃ m : ℕ, (separable_degree q hf) * (q ^ m) = f.nat_degree :=
+begin
+  cases (classical.some_spec hf).2 with m hm,
+  use m,
+  let g := classical.some hf,
+  have deg_g : g.nat_degree = separable_degree q hf, by refl,
+  rw [←deg_g, ←hm, nat_degree_expand (q^m) g, mul_comm],
+end
+
+/-- The separable degree divides the degree. -/
+lemma separable_degree_dvd_degree :
+  (separable_degree q hf) ∣ f.nat_degree :=
+let ⟨a, ha⟩ := separable_degree_dvd_degree' q hf in dvd.intro (q ^ a) ha
+
+/-- In exponential characteristic one, the separable degree equals the degree. -/
+lemma separable_degree_eq_degree {f : polynomial F}
+  (hf : has_separable_contraction 1 f) : (separable_degree 1 hf) = f.nat_degree :=
+exists.elim (separable_degree_dvd_degree' 1 hf) (λ a, λ ha, by rw [←ha, one_pow a, mul_one])
+
+lemma exists_eq_add_nonneg_of_lt (m m' : ℕ) (h : m < m') : ∃ s : ℕ, 0 < s ∧ m' = m + s :=
+let ⟨k, hk⟩ := nat.exists_eq_add_of_lt h in ⟨k.succ, k.zero_lt_succ, hk⟩
+
+end comm_semiring
+
+section field
+
+variables {F : Type} [field F]
+  (q : ℕ)
+  {f : polynomial F} (hf : has_separable_contraction q f)
+
+/-- Every irreducible polynomial can be contracted to a separable polynomial.
+https://stacks.math.columbia.edu/tag/09H0 -/
+lemma irreducible_has_separable_contraction (q : ℕ) [hF : exp_char F q]
+  (f : polynomial F) [irred : irreducible f] (fn : f ≠ 0) : has_separable_contraction q f :=
+begin
+  casesI hF,
+  { use f,
+    exact ⟨irreducible.separable irred, ⟨0, by rw [pow_zero, expand_one]⟩⟩ },
+  { haveI qp : fact (nat.prime q) := fact_iff.mpr hF_hprime,
+    rcases exists_separable_of_irreducible q irred fn with ⟨n, g, hg⟩,
+    use g,
+    exact ⟨hg.1, ⟨n, hg.2⟩⟩, }
```

Comment:
You can split further with `rcases` rather than using `1` and `2` here:
```suggestion
  { haveI qp : fact (nat.prime q) := ⟨hF_hprime⟩,
    rcases exists_separable_of_irreducible q irred fn with ⟨n, g, hgs, hge⟩,
    exact ⟨g, hgs, n, hge⟩, }
```

Reply:
Done.